src/storage/src/hummock/iterator/merge_inner.rs

```diff
@@ -202,41 +205,43 @@ impl<I: HummockIterator> MergeIteratorNext for OrderedMergeIteratorInner<I> {
 
     fn next_inner(&mut self) -> Self::HummockResultFuture<'_> {
         async {
-            let top_node = self.heap.pop().expect("no inner iter");
-            let mut popped_nodes = vec![];
-
-            // Take all nodes with the same current key as the top_node out of the heap.
-            while let Some(next_node) = self.heap.peek_mut() {
-                match top_node.iter.key().cmp(&next_node.iter.key()) {
-                    Ordering::Equal => {
-                        popped_nodes.push(PeekMut::pop(next_node));
-                    }
-                    _ => break,
+            let top_key = {
+                let top_key = self.heap.peek().expect("no inner iter").iter.key();
+                self.last_table_key.clear();
+                self.last_table_key
+                    .extend_from_slice(top_key.user_key.table_key.0);
+                FullKey {
+                    user_key: UserKey {
+                        table_id: top_key.user_key.table_id,
+                        table_key: TableKey(self.last_table_key.as_slice()),
+                    },
+                    epoch: top_key.epoch,
                 }
-            }
-
-            popped_nodes.push(top_node);
-
-            // WARNING: within scope of BinaryHeap::PeekMut, we must carefully handle all places of
-            // return. Once the iterator enters an invalid state, we should remove it from heap
-            // before returning.
-
-            // Put the popped nodes back to the heap if valid or unused_iters if invalid.
-            for mut node in popped_nodes {
-                match node.iter.next().await {
-                    Ok(_) => {}
-                    Err(e) => {
-                        // If the iterator returns error, we should clear the heap, so that this
-                        // iterator becomes invalid.
+            };
+            loop {
+                // We don't use `while let` or `if let` here because the current borrow checker is
```

Comment:
Does `let Some(xx) = xxx else { break }` work here?

Reply:
Thanks. It does work. I have fixed it.